# Practical Machine Learning for Physicists
## Coursework C -- Part 1


### Task 1:
Design, implement and test a neural network utilising a single convolutional layer (use as many other non convolutional layers as you need) to classify the MNIST handwritten digits. What is the maximum test accuracry you can achieve using a single convolutional layer?

### Task 2:
Design, implement and test a neural network utitlising multiple convolutional layers (again use as many other non convolutinal laters as you need) to classify the MNIST handwritten digits. What is the maximum test accuracry you can achieve using as many convolutional layers as you like?

#### Practicalities
You should use this notebook for your work and upload it to Moodle. You are expected to use TensorFlow and Keras to complete these takss. The notebook should be self-contained and able to be executed if necessary. Marks will be awarded for (roughly equally weighted):
- Overall notebook clarity (both in terms of good coding practice and coherent discussion)
- Network performance (how well does your classifier do?)
- Network efficiency (how does your network compare to the optimum networks for this task?)
- Network training (do you do a good job of traning your network?)


# Overview

Train a model with Conventional layer to classify the MNIST handwritten digits. The aim is to achieve the maximum test accuracy using both single convolutional layer and multiple convolutional layers

## Import necessary library

In [ ]:
## importing the necessary library


import matplotlib.pyplot as plt
import numpy as np
import math

# TensorFlow and tf.keras
import tensorflow as tf
from tensorflow import keras

import matplotlib.style #Some style nonsense
import matplotlib as mpl #Some more style nonsense


#Set default figure size
#mpl.rcParams['figure.figsize'] = [12.0, 8.0] #Inches... of course it is inches
mpl.rcParams["legend.frameon"] = False
mpl.rcParams['figure.dpi']=200 # dots per inch

#Useful for debugging problems
print(tf.__version__)

2.15.0


In [ ]:
## Load the MNIST dataset
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

print(train_images.shape)
print(train_labels.shape)
print(test_images.shape)
print(test_labels.shape)

11490434/11490434 [==============================] - 0s 0us/step
(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


In [ ]:
# normalise the image's pixel

train_images=train_images/255.0
test_images=test_images/255.0

## Neutral Network Containing a single Convolutional neutral layer

In this section of the notebook, we design a model with a single convolutional neutral layer which is then followed by a down pooling and then flatten to go through the dense hidden dense layer with 40 neurons. Finally, the model return 10 outputs (with probably for each of them) which corresponds to possible digits in the MNIST data.

In [ ]:
# Design the neutral network layers

model = keras.Sequential([
    keras.layers.Conv2D(32,kernel_size=(3,3),activation='relu', input_shape=(28,28,1)),
    keras.layers.AveragePooling2D(pool_size=(3,3),padding='same'),
    keras.layers.Flatten(),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dense(10,activation='softmax')
])

model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True), optimizer=tf.keras.optimizers.SGD(lr=1.0),
              metrics=['accuracy'])
model.summary()



Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 average_pooling2d (Average  (None, 9, 9, 32)          0         
 Pooling2D)                                                      
                                                                 
 flatten (Flatten)           (None, 2592)              0         
                                                                 
 dense (Dense)               (None, 30)                77790     
                                                                 
 dense_1 (Dense)             (None, 10)                310       
                                                                 
Total params: 78420 (306.33 KB)
Trainable params: 78420 (306.33 KB)
Non-trainable params: 0 (0.00 Byte)
__________________

## Train the model

In [ ]:
history=model.fit(train_images, train_labels,batch_size=128, epochs=30)

Epoch 1/30


/usr/local/lib/python3.10/dist-packages/keras/src/backend.py:5727: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


469/469 [==============================] - 19s 39ms/step - loss: 1.5035 - accuracy: 0.6035
Epoch 2/30
469/469 [==============================] - 16s 35ms/step - loss: 0.4949 - accuracy: 0.8690
Epoch 3/30
469/469 [==============================] - 16s 35ms/step - loss: 0.3751 - accuracy: 0.8947
Epoch 4/30
469/469 [==============================] - 17s 35ms/step - loss: 0.3356 - accuracy: 0.9039
Epoch 5/30
469/469 [==============================] - 16s 34ms/step - loss: 0.3129 - accuracy: 0.9092
Epoch 6/30
469/469 [==============================] - 17s 35ms/step - loss: 0.2968 - accuracy: 0.9139
Epoch 7/30
469/469 [==============================] - 16s 33ms/step - loss: 0.2843 - accuracy: 0.9171
Epoch 8/30
469/469 [==============================] - 16s 35ms/step - loss: 0.2729 - accuracy: 0.9198
Epoch 9/30
469/469 [==============================] - 16s 34ms/step - loss: 0.2638 - accuracy: 0.9231
Epoch 10/30
469/469 [==============================] - 16s 33ms/step - loss: 0.2547 - accurac

## Testing the model using the MNIST test sample


In [ ]:
test_loss,test_acc=model.evaluate(test_images,test_labels,verbose=2)
print("The accuracy of the trained model is ", test_acc *100, "%" )

313/313 - 2s - loss: 0.1662 - accuracy: 0.9510 - 2s/epoch - 6ms/step
The accuracy of the trained model is  95.09999752044678 %


## Neutral Network Containing Multiple Convolution Neutral layers

similar idea to the previous single convolution neutral network but with additional convolutional neutral layer being added.

In [ ]:
# Design a new multi-convolution neutral network

model2 = keras.Sequential([
    keras.layers.Conv2D(32,kernel_size=(3,3),activation='relu', input_shape=(28,28,1),padding='same'),
    keras.layers.AveragePooling2D(pool_size=(3,3),padding='same'),
    keras.layers.Conv2D(64,kernel_size=(3,3),activation='relu',padding='same'),
    keras.layers.AveragePooling2D(pool_size=(3,3),padding='same'),
    keras.layers.Flatten(),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dense(10,activation='softmax')
])

model2.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True), optimizer=tf.keras.optimizers.SGD(lr=1.0),
              metrics=['accuracy'])
model2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_1 (Conv2D)           (None, 28, 28, 32)        320       
                                                                 
 average_pooling2d_1 (Avera  (None, 10, 10, 32)        0         
 gePooling2D)                                                    
                                                                 
 conv2d_2 (Conv2D)           (None, 10, 10, 64)        18496     
                                                                 
 average_pooling2d_2 (Avera  (None, 4, 4, 64)          0         
 gePooling2D)                                                    
                                                                 
 flatten_1 (Flatten)         (None, 1024)              0         
                                                                 
 dense_2 (Dense)             (None, 30)               

## Train the model

In [ ]:
history2=model2.fit(train_images, train_labels,batch_size=128, epochs=30)

Epoch 1/30
469/469 [==============================] - 37s 78ms/step - loss: 2.2862 - accuracy: 0.2414
Epoch 2/30
469/469 [==============================] - 37s 79ms/step - loss: 2.1704 - accuracy: 0.4527
Epoch 3/30
469/469 [==============================] - 36s 77ms/step - loss: 1.3713 - accuracy: 0.6640
Epoch 4/30
469/469 [==============================] - 37s 79ms/step - loss: 0.7218 - accuracy: 0.7872
Epoch 5/30
469/469 [==============================] - 37s 78ms/step - loss: 0.5784 - accuracy: 0.8250
Epoch 6/30
469/469 [==============================] - 36s 77ms/step - loss: 0.5003 - accuracy: 0.8487
Epoch 7/30
469/469 [==============================] - 36s 78ms/step - loss: 0.4425 - accuracy: 0.8659
Epoch 8/30
469/469 [==============================] - 36s 78ms/step - loss: 0.3969 - accuracy: 0.8806
Epoch 9/30
469/469 [==============================] - 36s 76ms/step - loss: 0.3609 - accuracy: 0.8926
Epoch 10/30
469/469 [==============================] - 37s 78ms/step - loss: 0.331

## Test the model using the mnist test sample

In [ ]:
test2_loss,test2_acc=model2.evaluate(test_images,test_labels,verbose=2)
print("The accuracy of the trained model is ", test2_acc *100, "%" )

/usr/local/lib/python3.10/dist-packages/keras/src/backend.py:5727: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


313/313 - 3s - loss: 0.1221 - accuracy: 0.9607 - 3s/epoch - 9ms/step
The accuracy of the trained model is  96.069997549057 %


# Comment on the performance of both model
By using multiple convolutional layer, we have achieve a better accuracy of 96% compare to a single convolutional layer with accuracy of 95%. This is because by using multiple convolutional layer we can further reduce the chance of overfitting the model to the training data.